### PR TITLE
Merge volume flags implementation

### PR DESCRIPTION
**Pull request by mheon**
Merge the implementations of `--mount`, `--volume`, `--volumes-from`, `--tmpfs`, the mount portion of `--init`, and image volumes together into one unified API. Conflicting volumes and superceding volume priorities should be handled correctly here.

As part of this, move all volume-handling code into a single file in `pkg/spec`, storage.go.

Also, add a fresh API entrypoint for generating everything necessary to create a Libpod container out of a CreateConfig, and start making many things in `pkg/spec` private. A single unified entrypoint lets us do extensive refactoring of the package and still not break any users, because the API remains the same.

There are still a lot of WIPs left in the code, but I will hopefully hit them all tomorrow.

### CODE REVIEW

**Comment by mheon** [comment]
Note to self: this needs testing, especially around where the volume flags interact.

**Comment by mheon** [comment]
Goconst and I are having some disagreements here, evidently

**Comment by mheon** [comment]
Alright, existing unit tests are moved over. I'm going to start writing new ones, plus some integration tests, then this should be good to merge.

**Comment by mheon** [comment]
Alright, this PR is unwieldly enough as it is.

I'm going to strip the WIP off and stop adding code now, with the exception of fixing tests.

**Comment by mheon** [comment]
@vrothberg @rhatdan @baude @TomSweeneyRedHat PTAL. I've created a real monster here, but the end result seems a lot more understandable and maintainable than existing code.

File: pkg/spec/containerconfig.go [hunk 1]
```diff
@@ -3,11 +3,18 @@ package createconfig
...
 	spec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
 )
 
 // MakeContainerConfig generates all configuration necessary to start a
 // container with libpod from a completed CreateConfig struct.
 func (config *CreateConfig) MakeContainerConfig(runtime *libpod.Runtime, pod *libpod.Pod) (*spec.Spec, []libpod.CtrCreateOption, error) {
+	if config.Pod != "" && pod == nil {
+		return nil, nil, errors.Wrapf(libpod.ErrInvalidArg, "pod was specified but no pod passed")
+	} else if config.Pod == "" && pod != nil {
```
Comment (rhatdan) [inline pkg/spec/containerconfig.go]: Nit: No need for the else here.

File: pkg/spec/containerconfig.go [hunk 2]
```diff
@@ -25,5 +36,26 @@ func (config *CreateConfig) MakeContainerConfig(runtime *libpod.Runtime, pod *li
...
+// and named volumes to add to the container.
+// Handles --volumes-from, --volumes, and --mounts flags.
+func (config *CreateConfig) parseVolumes(runtime *libpod.Runtime) ([]spec.Mount, []*libpod.ContainerNamedVolume, error) {
+	// Start with --volumes-from
+	mounts, namedVolumes, err := config.getVolumesFrom(runtime)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Next --volumes flag
```
Comment (rhatdan) [inline pkg/spec/containerconfig.go]: Do we want to leave these comments?

File: pkg/spec/createconfig.go [hunk 3]
```diff
@@ -262,22 +262,40 @@ func (c *CreateConfig) GetVolumeMounts(specMounts []spec.Mount) ([]spec.Mount, e
...
+	// TODO: Should we unmount subtree mounts? E.g., if /tmp/ is mounted by
+	// one mount, and we already have /tmp/a and /tmp/b, should we remove
+	// the /tmp/a and /tmp/b mounts in favor of the more general /tmp?
+	finalMounts := make(map[string]spec.Mount)
+	finalNamedVolumes := make(map[string]*libpod.ContainerNamedVolume)
 
 	for _, vol := range c.VolumesFrom {
-		options := ""
+		options := []string{}
+		// TODO: now that container names may contain colons this is no
```
Comment (rhatdan) [inline pkg/spec/createconfig.go]: I thought we were not allowing this.

File: pkg/spec/createconfig.go [hunk 4]
```diff
@@ -322,11 +344,15 @@ func (c *CreateConfig) GetVolumesFrom(runtime *libpod.Runtime) error {
 			if _, exists := userVolumes[namedVol.Dest]; exists {
 				userVolumes[namedVol.Dest] = true
 			}
-			localOptions := options
-			if localOptions == "" {
-				localOptions = strings.Join(namedVol.Options, ",")
+
+			if len(options) != 0 {
```
Comment (rhatdan) [inline pkg/spec/createconfig.go]: Would namedVol.Options ever be non-zero at this point, If not then why check the length of options.

File: pkg/spec/spec.go [hunk 5]
```diff
@@ -396,9 +396,9 @@ func (config *CreateConfig) createConfigToOCISpec(runtime *libpod.Runtime) (*spe
 	}
 
 	// BIND MOUNTS
-	if err := config.GetVolumesFrom(runtime); err != nil {
-		return nil, nil, errors.Wrap(err, "error getting volume mounts from --volumes-from flag")
-	}
+	// if err := config.GetVolumesFrom(runtime); err != nil {
```
Comment (rhatdan) [inline pkg/spec/spec.go]: I really dislike commented out code. Since it seems to never get removed.

File: pkg/spec/spec.go [hunk 6]
```diff
@@ -394,16 +394,16 @@ func (config *CreateConfig) createConfigToOCISpec(runtime *libpod.Runtime) (*spe
 	// 	return nil, nil, errors.Wrap(err, "error getting volume mounts from --volumes-from flag")
 	// }
 
-	volumeMounts, err := config.GetVolumeMounts(configSpec.Mounts)
-	if err != nil {
-		return nil, nil, errors.Wrapf(err, "error getting volume mounts")
-	}
+	// volumeMounts, err := config.GetVolumeMounts(configSpec.Mounts)
```
Comment (rhatdan) [inline pkg/spec/spec.go]: More Comments?

File: pkg/spec/storage.go [hunk 7]
```diff
@@ -61,6 +66,8 @@ func (config *CreateConfig) parseVolumes(runtime *libpod.Runtime) ([]spec.Mount,
 	// Unify --mount, --volume, --tmpfs mounts
 	unifiedMounts := make(map[string]spec.Mount)
 
+	// Add image volumes
```
Comment (rhatdan) [inline pkg/spec/storage.go]: Are you missing part of the patch set?

File: pkg/spec/createconfig.go (hunk 3 above)
Comment (mheon) [reply pkg/spec/createconfig.go]: Removed comment in newer version

File: pkg/spec/spec.go (hunk 5 above)
Comment (mheon) [reply pkg/spec/spec.go]: Nuked it in later versions, was there as a reference to where I removed things

File: pkg/spec/storage.go (hunk 7 above)
Comment (mheon) [reply pkg/spec/storage.go]: I think you're on an older version of the patch - the rest is addressed in the latest commit

Comment (rhatdan) [reply pkg/spec/storage.go]: Yes I think I was reviewing this when you force pushed.  
Anyways no way to do a great job on reviewing this, so hope tests catch any failures.

**Comment by baude** [comment]
LGTM.

**Comment by mheon** [comment]
Alright, tests should go green now.
@rhatdan @vrothberg @giuseppe @TomSweeneyRedHat anyone want to give this a second LGTM?

**Comment by mheon** [comment]
It's green
@rhatdan @baude @TomSweeneyRedHat @vrothberg Someone want to push this one over the finish line?

**Comment by TomSweeneyRedHat** [comment]
LGTM, I'll push and we can adjust if necessary.

**Pull request merged**